File: backend/app/routers/alerts.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import and_
from ..database import get_db
from ..models import AlertRule, AlertEvent, Cluster, User
from ..auth import get_current_user
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
import json
from ..core.logging import get_logger

router = APIRouter()
logger = get_logger(__name__)


class ThresholdConfig(BaseModel):
    cpu_percent: Optional[float] = None
    memory_percent: Optional[float] = None
    restart_count: Optional[int] = None
    time_window_seconds: Optional[int] = None


class AlertRuleCreate(BaseModel):
    name: str
    cluster_id: int
    rule_type: str
    severity: str = "warning"
    enabled: bool = True
    threshold_config: ThresholdConfig
    notification_channels: Optional[List[str]] = None


class AlertRuleUpdate(BaseModel):
    name: Optional[str] = None
    severity: Optional[str] = None
    enabled: Optional[bool] = None
    threshold_config: Optional[ThresholdConfig] = None
    notification_channels: Optional[List[str]] = None


class AlertRuleResponse(BaseModel):
    id: int
    name: str
    cluster_id: int
    cluster_name: str
    rule_type: str
    severity: str
    enabled: bool
    threshold_config: Dict[str, Any]
    notification_channels: Optional[List[str]]
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class AlertEventResponse(BaseModel):
    id: int
    rule_id: int
    rule_name: str
    cluster_id: int
    cluster_name: str
    resource_type: str
    resource_name: str
    namespace: Optional[str]
    severity: str
    message: str
    details: Optional[Dict[str, Any]]
    status: str
    first_triggered_at: datetime
    last_triggered_at: datetime
    resolved_at: Optional[datetime]
    notification_sent: bool

    class Config:
        from_attributes = True


@router.post("/rules", response_model=AlertRuleResponse)
async def create_alert_rule(
    rule: AlertRuleCreate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """创建告警规则"""
    try:
        cluster = db.query(Cluster).filter(Cluster.id == rule.cluster_id).first()
        if not cluster:
            raise HTTPException(status_code=404, detail="集群不存在")

        db_rule = AlertRule(
            name=rule.name,
            cluster_id=rule.cluster_id,
            rule_type=rule.rule_type,
            severity=rule.severity,
            enabled=rule.enabled,
            threshold_config=json.dumps(rule.threshold_config.model_dump(exclude_none=True)),
            notification_channels=json.dumps(rule.notification_channels) if rule.notification_channels else None,
            created_by=current_user["id"]
        )
        db.add(db_rule)
        db.commit()
        db.refresh(db_rule)

        return AlertRuleResponse(
            id=db_rule.id,
            name=db_rule.name,
            cluster_id=db_rule.cluster_id,
            cluster_name=cluster.name,
            rule_type=db_rule.rule_type,
            severity=db_rule.severity,
            enabled=db_rule.enabled,
            threshold_config=json.loads(db_rule.threshold_config),
            notification_channels=json.loads(db_rule.notification_channels) if db_rule.notification_channels else None,
            created_by=db_rule.created_by,
            created_at=db_rule.created_at,
            updated_at=db_rule.updated_at
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("创建告警规则失败: %s", e)
        db.rollback()
        raise HTTPException(status_code=500, detail=f"创建告警规则失败: {str(e)}")


@router.get("/rules", response_model=List[AlertRuleResponse])
async def list_alert_rules(
    cluster_id: Optional[int] = Query(None),
    enabled: Optional[bool] = Query(None),
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """获取告警规则列表"""
    try:
        query = db.query(AlertRule)

        if cluster_id is not None:
            query = query.filter(AlertRule.cluster_id == cluster_id)
        if enabled is not None:
            query = query.filter(AlertRule.enabled == enabled)

        rules = query.all()

        result = []
        for rule in rules:
            cluster = db.query(Cluster).filter(Cluster.id == rule.cluster_id).first()
            result.append(AlertRuleResponse(
                id=rule.id,
                name=rule.name,
                cluster_id=rule.cluster_id,
                cluster_name=cluster.name if cluster else "Unknown",
                rule_type=rule.rule_type,
                severity=rule.severity,
                enabled=rule.enabled,
                threshold_config=json.loads(rule.threshold_config),
                notification_channels=json.loads(rule.notification_channels) if rule.notification_channels else None,
                created_by=rule.created_by,
                created_at=rule.created_at,
                updated_at=rule.updated_at
            ))

        return result

    except Exception as e:
        logger.exception("获取告警规则列表失败: %s", e)
        raise HTTPException(status_code=500, detail=f"获取告警规则列表失败: {str(e)}")


@router.get("/rules/{rule_id}", response_model=AlertRuleResponse)
async def get_alert_rule(
    rule_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """获取告警规则详情"""
    try:
        rule = db.query(AlertRule).filter(AlertRule.id == rule_id).first()
        if not rule:
            raise HTTPException(status_code=404, detail="告警规则不存在")

        cluster = db.query(Cluster).filter(Cluster.id == rule.cluster_id).first()

        return AlertRuleResponse(
            id=rule.id,
            name=rule.name,
            cluster_id=rule.cluster_id,
            cluster_name=cluster.name if cluster else "Unknown",
            rule_type=rule.rule_type,
            severity=rule.severity,
            enabled=rule.enabled,
            threshold_config=json.loads(rule.threshold_config),
            notification_channels=json.loads(rule.notification_channels) if rule.notification_channels else None,
            created_by=rule.created_by,
            created_at=rule.created_at,
            updated_at=rule.updated_at
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("获取告警规则详情失败: %s", e)
        raise HTTPException(status_code=500, detail=f"获取告警规则详情失败: {str(e)}")


@router.put("/rules/{rule_id}", response_model=AlertRuleResponse)
async def update_alert_rule(
    rule_id: int,
    rule_update: AlertRuleUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """更新告警规则"""
    try:
        rule = db.query(AlertRule).filter(AlertRule.id == rule_id).first()
        if not rule:
            raise HTTPException(status_code=404, detail="告警规则不存在")

        if rule_update.name is not None:
            rule.name = rule_update.name
        if rule_update.severity is not None:
            rule.severity = rule_update.severity
        if rule_update.enabled is not None:
            rule.enabled = rule_update.enabled
        if rule_update.threshold_config is not None:
            rule.threshold_config = json.dumps(rule_update.threshold_config.model_dump(exclude_none=True))
        if rule_update.notification_channels is not None:
            rule.notification_channels = json.dumps(rule_update.notification_channels)

        db.commit()
        db.refresh(rule)

        cluster = db.query(Cluster).filter(Cluster.id == rule.cluster_id).first()

        return AlertRuleResponse(
            id=rule.id,
            name=rule.name,
            cluster_id=rule.cluster_id,
            cluster_name=cluster.name if cluster else "Unknown",
            rule_type=rule.rule_type,
            severity=rule.severity,
            enabled=rule.enabled,
            threshold_config=json.loads(rule.threshold_config),
            notification_channels=json.loads(rule.notification_channels) if rule.notification_channels else None,
            created_by=rule.created_by,
            created_at=rule.created_at,
            updated_at=rule.updated_at
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("更新告警规则失败: %s", e)
        db.rollback()
        raise HTTPException(status_code=500, detail=f"更新告警规则失败: {str(e)}")


@router.delete("/rules/{rule_id}")
async def delete_alert_rule(
    rule_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """删除告警规则"""
    try:
        rule = db.query(AlertRule).filter(AlertRule.id == rule_id).first()
        if not rule:
            raise HTTPException(status_code=404, detail="告警规则不存在")

        db.delete(rule)
        db.commit()

        return {"message": "告警规则已删除"}

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("删除告警规则失败: %s", e)
        db.rollback()
        raise HTTPException(status_code=500, detail=f"删除告警规则失败: {str(e)}")


@router.get("/events", response_model=List[AlertEventResponse])
async def list_alert_events(
    cluster_id: Optional[int] = Query(None),
    status: Optional[str] = Query(None),
    severity: Optional[str] = Query(None),
    limit: int = Query(100, le=1000),
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """获取告警事件列表"""
    try:
        query = db.query(AlertEvent)

        if cluster_id is not None:
            query = query.filter(AlertEvent.cluster_id == cluster_id)
        if status is not None:
            query = query.filter(AlertEvent.status == status)
        if severity is not None:
            query = query.filter(AlertEvent.severity == severity)

        events = query.order_by(AlertEvent.last_triggered_at.desc()).limit(limit).all()

        result = []
        for event in events:
            rule = db.query(AlertRule).filter(AlertRule.id == event.rule_id).first()
            cluster = db.query(Cluster).filter(Cluster.id == event.cluster_id).first()

            result.append(AlertEventResponse(
                id=event.id,
                rule_id=event.rule_id,
                rule_name=rule.name if rule else "Unknown",
                cluster_id=event.cluster_id,
                cluster_name=cluster.name if cluster else "Unknown",
                resource_type=event.resource_type,
                resource_name=event.resource_name,
                namespace=event.namespace,
                severity=event.severity,
                message=event.message,
                details=json.loads(event.details) if event.details else None,
                status=event.status,
                first_triggered_at=event.first_triggered_at,
                last_triggered_at=event.last_triggered_at,
                resolved_at=event.resolved_at,
                notification_sent=event.notification_sent
            ))

        return result

    except Exception as e:
        logger.exception("获取告警事件列表失败: %s", e)
        raise HTTPException(status_code=500, detail=f"获取告警事件列表失败: {str(e)}")


@router.post("/events/{event_id}/resolve")
async def resolve_alert_event(
    event_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """标记告警事件为已解决"""
    try:
        event = db.query(AlertEvent).filter(AlertEvent.id == event_id).first()
        if not event:
            raise HTTPException(status_code=404, detail="告警事件不存在")

        event.status = "resolved"
        event.resolved_at = datetime.now()
        db.commit()

        return {"message": "告警事件已标记为已解决"}

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("解决告警事件失败: %s", e)
        db.rollback()
        raise HTTPException(status_code=500, detail=f"解决告警事件失败: {str(e)}")


@router.get("/stats")
async def get_alert_stats(
    cluster_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user)
):
    """获取告警统计信息"""
    try:
        query = db.query(AlertEvent)
        if cluster_id is not None:
            query = query.filter(AlertEvent.cluster_id == cluster_id)

        total = query.count()
        firing = query.filter(AlertEvent.status == "firing").count()
        resolved = query.filter(AlertEvent.status == "resolved").count()

        critical = query.filter(and_(AlertEvent.severity == "critical", AlertEvent.status == "firing")).count()
        warning = query.filter(and_(AlertEvent.severity == "warning", AlertEvent.status == "firing")).count()
        info = query.filter(and_(AlertEvent.severity == "info", AlertEvent.status == "firing")).count()

        return {
            "total": total,
            "firing": firing,
            "resolved": resolved,
            "by_severity": {
                "critical": critical,
                "warning": warning,
                "info": info
            }
        }

    except Exception as e:
        logger.exception("获取告警统计信息失败: %s", e)
        raise HTTPException(status_code=500, detail=f"获取告警统计信息失败: {str(e)}")

File: backend/internal/httpapi/handlers/alerts.go
package handlers

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"time"

	"canvas/backend/internal/httpapi/middleware"
	"canvas/backend/internal/httpapi/response"
	"canvas/backend/internal/models"
	"github.com/go-chi/chi/v5"
	"gorm.io/gorm"
)

type AlertHandler struct {
	DB *gorm.DB
}

func NewAlertHandler(db *gorm.DB) *AlertHandler {
	return &AlertHandler{DB: db}
}

type alertRuleCreateRequest struct {
	Name                 string                 `json:"name"`
	ClusterID            uint                   `json:"cluster_id"`
	RuleType             string                 `json:"rule_type"`
	Severity             string                 `json:"severity"`
	Enabled              *bool                  `json:"enabled"`
	ThresholdConfig      map[string]interface{} `json:"threshold_config"`
	NotificationChannels []string               `json:"notification_channels"`
}

type alertRuleUpdateRequest struct {
	Name                 *string                `json:"name"`
	Severity             *string                `json:"severity"`
	Enabled              *bool                  `json:"enabled"`
	ThresholdConfig      map[string]interface{} `json:"threshold_config"`
	NotificationChannels *[]string              `json:"notification_channels"`
}

func (h *AlertHandler) CreateRule(w http.ResponseWriter, r *http.Request) {
	currentUser, ok := middleware.CurrentUser(r)
	if !ok {
		response.Error(w, r, http.StatusUnauthorized, "authentication required")
		return
	}

	var req alertRuleCreateRequest
	if err := response.DecodeJSON(r, &req); err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid request body")
		return
	}

	req.Name = strings.TrimSpace(req.Name)
	req.RuleType = strings.TrimSpace(req.RuleType)
	req.Severity = strings.TrimSpace(req.Severity)
	if req.Severity == "" {
		req.Severity = "warning"
	}
	if req.Name == "" || req.RuleType == "" || req.ClusterID == 0 {
		response.Error(w, r, http.StatusBadRequest, "name, cluster_id and rule_type are required")
		return
	}
	if req.ThresholdConfig == nil {
		req.ThresholdConfig = map[string]interface{}{}
	}

	var cluster models.Cluster
	if err := h.DB.First(&cluster, req.ClusterID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			response.Error(w, r, http.StatusNotFound, "集群不存在")
			return
		}
		response.Error(w, r, http.StatusInternalServerError, "database error")
		return
	}

	thresholdJSON, err := json.Marshal(req.ThresholdConfig)
	if err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid threshold_config")
		return
	}
	if len(thresholdJSON) == 0 {
		thresholdJSON = []byte("{}")
	}

	var channelsJSON *string
	if req.NotificationChannels != nil {
		raw, err := json.Marshal(req.NotificationChannels)
		if err != nil {
			response.Error(w, r, http.StatusBadRequest, "invalid notification_channels")
			return
		}
		text := string(raw)
		channelsJSON = &text
	}

	enabled := true
	if req.Enabled != nil {
		enabled = *req.Enabled
	}

	rule := models.AlertRule{
		Name:                 req.Name,
		ClusterID:            req.ClusterID,
		RuleType:             req.RuleType,
		Severity:             req.Severity,
		Enabled:              enabled,
		ThresholdConfig:      string(thresholdJSON),
		NotificationChannels: channelsJSON,
		CreatedBy:            currentUser.ID,
	}

	if err := h.DB.Create(&rule).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "创建告警规则失败")
		return
	}

	response.Success(w, r, http.StatusOK, mapAlertRule(rule, cluster.Name))
}

func (h *AlertHandler) ListRules(w http.ResponseWriter, r *http.Request) {
	query := h.DB.Model(&models.AlertRule{})

	clusterIDRaw := strings.TrimSpace(r.URL.Query().Get("cluster_id"))
	if clusterIDRaw != "" {
		clusterID, err := parsePathUintParam(clusterIDRaw)
		if err != nil || clusterID == 0 {
			response.Error(w, r, http.StatusBadRequest, "invalid cluster_id")
			return
		}
		query = query.Where("cluster_id = ?", clusterID)
	}

	enabledRaw := strings.TrimSpace(r.URL.Query().Get("enabled"))
	if enabledRaw != "" {
		enabled, err := strconv.ParseBool(enabledRaw)
		if err != nil {
			response.Error(w, r, http.StatusBadRequest, "invalid enabled")
			return
		}
		query = query.Where("enabled = ?", enabled)
	}

	var rules []models.AlertRule
	if err := query.Order("id ASC").Find(&rules).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警规则列表失败")
		return
	}

	clusterNames := h.loadClusterNamesForRules(rules)
	items := make([]map[string]interface{}, 0, len(rules))
	for _, rule := range rules {
		items = append(items, mapAlertRule(rule, clusterNames[rule.ClusterID]))
	}

	response.Success(w, r, http.StatusOK, items)
}

func (h *AlertHandler) GetRule(w http.ResponseWriter, r *http.Request) {
	ruleID, err := parsePathUint(chi.URLParam(r, "ruleID"))
	if err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid rule id")
		return
	}

	var rule models.AlertRule
	if err := h.DB.First(&rule, ruleID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			response.Error(w, r, http.StatusNotFound, "告警规则不存在")
			return
		}
		response.Error(w, r, http.StatusInternalServerError, "database error")
		return
	}

	clusterName := h.getClusterName(rule.ClusterID)
	response.Success(w, r, http.StatusOK, mapAlertRule(rule, clusterName))
}

func (h *AlertHandler) UpdateRule(w http.ResponseWriter, r *http.Request) {
	ruleID, err := parsePathUint(chi.URLParam(r, "ruleID"))
	if err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid rule id")
		return
	}

	var req alertRuleUpdateRequest
	if err := response.DecodeJSON(r, &req); err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid request body")
		return
	}

	var rule models.AlertRule
	if err := h.DB.First(&rule, ruleID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			response.Error(w, r, http.StatusNotFound, "告警规则不存在")
			return
		}
		response.Error(w, r, http.StatusInternalServerError, "database error")
		return
	}

	if req.Name != nil {
		trimmed := strings.TrimSpace(*req.Name)
		if trimmed == "" {
			response.Error(w, r, http.StatusBadRequest, "name cannot be empty")
			return
		}
		rule.Name = trimmed
	}

	if req.Severity != nil {
		trimmed := strings.TrimSpace(*req.Severity)
		if trimmed != "" {
			rule.Severity = trimmed
		}
	}

	if req.Enabled != nil {
		rule.Enabled = *req.Enabled
	}

	if req.ThresholdConfig != nil {
		thresholdJSON, err := json.Marshal(req.ThresholdConfig)
		if err != nil {
			response.Error(w, r, http.StatusBadRequest, "invalid threshold_config")
			return
		}
		rule.ThresholdConfig = string(thresholdJSON)
	}

	if req.NotificationChannels != nil {
		raw, err := json.Marshal(*req.NotificationChannels)
		if err != nil {
			response.Error(w, r, http.StatusBadRequest, "invalid notification_channels")
			return
		}
		text := string(raw)
		rule.NotificationChannels = &text
	}

	updatedAt := time.Now().UTC()
	rule.UpdatedAt = &updatedAt
	if err := h.DB.Save(&rule).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "更新告警规则失败")
		return
	}

	clusterName := h.getClusterName(rule.ClusterID)
	response.Success(w, r, http.StatusOK, mapAlertRule(rule, clusterName))
}

func (h *AlertHandler) DeleteRule(w http.ResponseWriter, r *http.Request) {
	ruleID, err := parsePathUint(chi.URLParam(r, "ruleID"))
	if err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid rule id")
		return
	}

	var rule models.AlertRule
	if err := h.DB.First(&rule, ruleID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			response.Error(w, r, http.StatusNotFound, "告警规则不存在")
			return
		}
		response.Error(w, r, http.StatusInternalServerError, "database error")
		return
	}

	if err := h.DB.Delete(&rule).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "删除告警规则失败")
		return
	}

	response.Success(w, r, http.StatusOK, map[string]string{"message": "告警规则已删除"})
}

func (h *AlertHandler) ListEvents(w http.ResponseWriter, r *http.Request) {
	query := h.DB.Model(&models.AlertEvent{})

	clusterIDRaw := strings.TrimSpace(r.URL.Query().Get("cluster_id"))
	if clusterIDRaw != "" {
		clusterID, err := parsePathUintParam(clusterIDRaw)
		if err != nil || clusterID == 0 {
			response.Error(w, r, http.StatusBadRequest, "invalid cluster_id")
			return
		}
		query = query.Where("cluster_id = ?", clusterID)
	}

	status := strings.TrimSpace(r.URL.Query().Get("status"))
	if status != "" {
		query = query.Where("status = ?", status)
	}
	severity := strings.TrimSpace(r.URL.Query().Get("severity"))
	if severity != "" {
		query = query.Where("severity = ?", severity)
	}

	limit := parseIntWithDefault(strings.TrimSpace(r.URL.Query().Get("limit")), 100)
	if limit <= 0 {
		limit = 100
	}
	if limit > 1000 {
		limit = 1000
	}

	var events []models.AlertEvent
	if err := query.Order("last_triggered_at DESC").Limit(limit).Find(&events).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警事件列表失败")
		return
	}

	ruleNames := h.loadRuleNamesForEvents(events)
	clusterNames := h.loadClusterNamesForEvents(events)

	items := make([]map[string]interface{}, 0, len(events))
	for _, event := range events {
		items = append(items, mapAlertEvent(event, ruleNames[event.RuleID], clusterNames[event.ClusterID]))
	}

	response.Success(w, r, http.StatusOK, items)
}

func (h *AlertHandler) ResolveEvent(w http.ResponseWriter, r *http.Request) {
	eventID, err := parsePathUint(chi.URLParam(r, "eventID"))
	if err != nil {
		response.Error(w, r, http.StatusBadRequest, "invalid event id")
		return
	}

	var event models.AlertEvent
	if err := h.DB.First(&event, eventID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			response.Error(w, r, http.StatusNotFound, "告警事件不存在")
			return
		}
		response.Error(w, r, http.StatusInternalServerError, "database error")
		return
	}

	now := time.Now().UTC()
	event.Status = "resolved"
	event.ResolvedAt = &now
	if err := h.DB.Save(&event).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "更新告警事件失败")
		return
	}

	response.Success(w, r, http.StatusOK, map[string]string{"message": "告警事件已标记为已解决"})
}

func (h *AlertHandler) Stats(w http.ResponseWriter, r *http.Request) {
	query := h.DB.Model(&models.AlertEvent{})

	clusterIDRaw := strings.TrimSpace(r.URL.Query().Get("cluster_id"))
	if clusterIDRaw != "" {
		clusterID, err := parsePathUintParam(clusterIDRaw)
		if err != nil || clusterID == 0 {
			response.Error(w, r, http.StatusBadRequest, "invalid cluster_id")
			return
		}
		query = query.Where("cluster_id = ?", clusterID)
	}

	var total int64
	if err := query.Count(&total).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}

	firingQuery := query.Session(&gorm.Session{}).Where("status = ?", "firing")
	resolvedQuery := query.Session(&gorm.Session{}).Where("status = ?", "resolved")

	var firing int64
	var resolved int64
	if err := firingQuery.Count(&firing).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}
	if err := resolvedQuery.Count(&resolved).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}

	var critical int64
	var warning int64
	var info int64
	if err := query.Session(&gorm.Session{}).Where("severity = ? AND status = ?", "critical", "firing").Count(&critical).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}
	if err := query.Session(&gorm.Session{}).Where("severity = ? AND status = ?", "warning", "firing").Count(&warning).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}
	if err := query.Session(&gorm.Session{}).Where("severity = ? AND status = ?", "info", "firing").Count(&info).Error; err != nil {
		response.Error(w, r, http.StatusInternalServerError, "获取告警统计失败")
		return
	}

	response.Success(w, r, http.StatusOK, map[string]interface{}{
		"total":    total,
		"firing":   firing,
		"resolved": resolved,
		"by_severity": map[string]interface{}{
			"critical": critical,
			"warning":  warning,
			"info":     info,
		},
	})
}

func (h *AlertHandler) loadClusterNamesForRules(rules []models.AlertRule) map[uint]string {
	ids := make([]uint, 0)
	seen := map[uint]struct{}{}
	for _, rule := range rules {
		if _, ok := seen[rule.ClusterID]; ok {
			continue
		}
		seen[rule.ClusterID] = struct{}{}
		ids = append(ids, rule.ClusterID)
	}
	return h.loadClusterNames(ids)
}

func (h *AlertHandler) loadClusterNamesForEvents(events []models.AlertEvent) map[uint]string {
	ids := make([]uint, 0)
	seen := map[uint]struct{}{}
	for _, event := range events {
		if _, ok := seen[event.ClusterID]; ok {
			continue
		}
		seen[event.ClusterID] = struct{}{}
		ids = append(ids, event.ClusterID)
	}
	return h.loadClusterNames(ids)
}

func (h *AlertHandler) loadClusterNames(ids []uint) map[uint]string {
	if len(ids) == 0 {
		return map[uint]string{}
	}
	var clusters []models.Cluster
	if err := h.DB.Where("id IN ?", ids).Find(&clusters).Error; err != nil {
		return map[uint]string{}
	}
	items := make(map[uint]string, len(clusters))
	for _, cluster := range clusters {
		items[cluster.ID] = cluster.Name
	}
	return items
}

func (h *AlertHandler) loadRuleNamesForEvents(events []models.AlertEvent) map[uint]string {
	ids := make([]uint, 0)
	seen := map[uint]struct{}{}
	for _, event := range events {
		if _, ok := seen[event.RuleID]; ok {
			continue
		}
		seen[event.RuleID] = struct{}{}
		ids = append(ids, event.RuleID)
	}

	if len(ids) == 0 {
		return map[uint]string{}
	}

	var rules []models.AlertRule
	if err := h.DB.Where("id IN ?", ids).Find(&rules).Error; err != nil {
		return map[uint]string{}
	}

	items := make(map[uint]string, len(rules))
	for _, rule := range rules {
		items[rule.ID] = rule.Name
	}
	return items
}

func (h *AlertHandler) getClusterName(clusterID uint) string {
	var cluster models.Cluster
	if err := h.DB.Select("name").First(&cluster, clusterID).Error; err != nil {
		return "Unknown"
	}
	if strings.TrimSpace(cluster.Name) == "" {
		return "Unknown"
	}
	return cluster.Name
}

func mapAlertRule(rule models.AlertRule, clusterName string) map[string]interface{} {
	threshold := map[string]interface{}{}
	if strings.TrimSpace(rule.ThresholdConfig) != "" {
		_ = json.Unmarshal([]byte(rule.ThresholdConfig), &threshold)
	}

	var channels []string
	if rule.NotificationChannels != nil && strings.TrimSpace(*rule.NotificationChannels) != "" {
		_ = json.Unmarshal([]byte(*rule.NotificationChannels), &channels)
	}

	updatedAt := rule.CreatedAt
	if rule.UpdatedAt != nil {
		updatedAt = *rule.UpdatedAt
	}

	return map[string]interface{}{
		"id":                    rule.ID,
		"name":                  rule.Name,
		"cluster_id":            rule.ClusterID,
		"cluster_name":          clusterName,
		"rule_type":             rule.RuleType,
		"severity":              rule.Severity,
		"enabled":               rule.Enabled,
		"threshold_config":      threshold,
		"notification_channels": channels,
		"created_by":            rule.CreatedBy,
		"created_at":            formatTimeRFC3339(rule.CreatedAt),
		"updated_at":            formatTimeRFC3339(updatedAt),
	}
}

func mapAlertEvent(event models.AlertEvent, ruleName string, clusterName string) map[string]interface{} {
	details := map[string]interface{}{}
	if event.Details != nil && strings.TrimSpace(*event.Details) != "" {
		_ = json.Unmarshal([]byte(*event.Details), &details)
	}

	var namespace interface{}
	if event.Namespace != nil && strings.TrimSpace(*event.Namespace) != "" {
		namespace = *event.Namespace
	}

	resolvedAt := interface{}(nil)
	if event.ResolvedAt != nil {
		resolvedAt = formatTimeRFC3339(*event.ResolvedAt)
	}

	return map[string]interface{}{
		"id":                 event.ID,
		"rule_id":            event.RuleID,
		"rule_name":          ruleName,
		"cluster_id":         event.ClusterID,
		"cluster_name":       clusterName,
		"resource_type":      event.ResourceType,
		"resource_name":      event.ResourceName,
		"namespace":          namespace,
		"severity":           event.Severity,
		"message":            event.Message,
		"details":            details,
		"status":             event.Status,
		"first_triggered_at": formatTimeRFC3339(event.FirstTriggeredAt),
		"last_triggered_at":  formatTimeRFC3339(event.LastTriggeredAt),
		"resolved_at":        resolvedAt,
		"notification_sent":  event.NotificationSent,
	}
}
